docs/arcadia-users-group/20231128-testing-datasets/lesson.md

```diff
@@ -0,0 +1,84 @@
+# Building test data sets
+
+In our previous lesson, we covered different types of tests.
+In the unit test section, we used five functions to test that our `mean()` function produced the output that we expected it to.
+For each of these functions, we started by defining our test data, `num_list`, a list of numbers for which we wanted to calculate the mean.
+
+These test data were small; each list only had five numbers in it.
+The "correct" answer to the test was clear with relatively little human effort (we could take a few seconds to calculate the mean of each list in our own heads).
+The creation of each test data set was encoded in the test itself.
+
+Below we discuss strategies for creating test data sets to help maintain some of these desirable qualities in real code bases.
+
+## Test data for functions and scripts
+
+### Creating data within the test itself
+
+As seen in our previous lesson and described above, one strategy is to create test data within the test using built-in functions or data structures.
+When possible, this is a great strategy.
+These test data are small and self-contained and the correct answer to the test is easy to figure out.
+Functions like `[]` (list creation), or `range()` in Python and `c()`, `rep()`, `seq()`, `data.frame()`, and `list()` in R can be used to create test data that are relevant to a function.
+Some data-creation functions might be non-deterministic, but setting the seed ensures that the same output are produced each time.
+
+### Testing image outputs
+
+Often times, a function or a script produces an image output.
+Images are trickier to test than some other outputs, but can still be used in unit tests.
+If a function produces the image and an image object is produced in your coding environment, you can sometimes check the metadata of the image to make sure it matches expected values.
+This includes dimensions, color profiles, or file formats.
+These types of tests have the built-in benefit of testing whether the function produces an image successfully on top of confirming attributes about the image.
+
+Even if you check metadata attributes about an image, the image itself might still be different than expected.
+If you want to test the output image more thoroughly, you can compare the output produced by a test to a reference image stored in your test data.
+Both checksum comparisons and pixel comparisons are useful in this scenario.
+
```

Comment:
````suggestion

We'll modify our function to generate an image -- a histogram of letter counts:

```python
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

def histogram(file_in, file_out):
    with open(file_in) as fi, open(file_out, 'w') as fo:
        counts = distribution(fi.read())
        df = pd.DataFrame(
            list(counts.items()), columns=['Character', 'Count']
        ).sort_values(by='Character')
        sns.barplot(data=df, x='Character', y='Count')
        plt.savefig(file_out)
```

Let's try it:
```
histogram('/usr/share/dict/propernames', '/tmp/propernames.png')
```

<img src="fig/propernames.png" alt="Character frequency in /usr/share/dict/propernames" width="50%"/>

To compare generated images to ground truth, we can use checksums as before, or we can visually compare them. For this, we'll use a Python image processing library which would require installation with `mamba install pillow`.

```python
from PIL import Image, ImageChops

def image_diff(file1, file2, file_out):
    """Compare file1 and file2; if different, store diff to file_out and assert."""
    img1 = Image.open(file1)
    img2 = Image.open(file2)
    diff = ImageChops.difference(img1, img2)
    if diff.getbbox():
        # Diff is the same as an empty image, i.e. `img1` and `img2` are identical.
        return True
    else:
        # Images differ, store a comparison of them.
        ImageChops.blend(img1, diff, alpha=0.5).save(file_out)
        assert False, f"Images differ, comparison in {file_out}"
```

For example, the histograms of the two samples discussed earlier -- the head of `/usr/share/dict/web2a` and a random sample from it - look relatively similar:

<img src="fig/web2a.head.png" alt="Character frequency in the head of /usr/share/dict/web2a" style="width: 50%; float: left;" />
<img src="fig/web2a.rand.png" alt="Character frequency in a random subset of /usr/share/dict/web2a" style="width: 50%; float: right;" />

But applying `image_diff()` highlights areas that have changed:

<div align="center">
    <img src="fig/web2a.diff.png" alt="Visual diff of top and random sampling from inputs" style="width: 50%" />
</div>

````